lib/ecto/query/planner.ex

```diff
@@ -1424,6 +1425,19 @@ defmodule Ecto.Query.Planner do
     end
   end
 
+  defp assert_valid_combinations!(query, operation) do
+    if query.combinations != [] do
+      has_bindings? =
+        Enum.any?(query.order_bys, fn order_by ->
+          Enum.any?(order_by.expr, &match?({_dir, {{:., [], [{:&, [], [0]}, :id]}, [], []}}, &1))
+        end)
+
+      if has_bindings? do
+        error! query, "`#{operation}` does not allow bindings in `order_by` when using `union` or `union_all`"
```

Comment:
We also have intersect, intersect_all, except and except_all.

Reply:
If you don't want to list all of them, you can do this:

```elixir
case query.combinations do
  [] -> :ok
  [{kind, _} | _] -> ...
end
```

and then use `kind` in the error message. We also don't need the `operation` here, because this is never valid, regardless of the operation. I would go with something like this:

```
error!, query, """
cannot use bindings in `order_by` when using `#{kind}`.

That's because the `order_by` applies to the whole `#{kind}` and not \
an individual query. If you really want to order the results, you can wrap \
the existing query in a subquery and then order it:

    query = query1 |> #{kind}(query2)
    from q in subquery(query), order_by: q.field